### ISSUE TITLE
mypy incorrectly missing return statement in presence of ExitStack

### ISSUE THREAD

**Issue by jll63** [python/mypy#10896]
**Bug Report**

`mypy` reports a missing return statement when `return` is inside a `with ExitStack()` block.

**To Reproduce**

Given the following file:

```python
# bug.py

import contextlib

def test() -> int:
    with contextlib.ExitStack():
        return 0
```

**Expected Behavior**

All paths return an object of the expected type. `mypy` should not report an error. 

**Actual Behavior**

```shell
$ mypy bug.py
bug.py:3: error: Missing return statement
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

```shell
$ mypy --version
mypy 0.910
```

**Comment by erictraut** [comment]
The `ExitStack.__exit__` method in `contextlib.pyi` returns a `bool` value, which tells a type checker that this context manager catches exceptions. If correct, then a type checker needs to assume that any statement within the context manager block could raise an exception, and control could resume after the context manager block. That is, it must assume that you could fall off the bottom of the function and implicitly return `None`.

So I think that mypy is doing the right thing here. If you think that `ExitStack.__exit__` is incorrectly annotated, please file a bug or a PR in typeshed.

**Comment by ethanhs** [comment]
Yeah, I think mypy is correct here. Unfortunately there isn't a better way to tell what paths will be taken.

**Comment by jll63** [comment]
> The `ExitStack.__exit__` method in `contextlib.pyi` returns a `bool` value, which tells a type checker that this context manager catches exceptions. If correct, then a type checker needs to assume that any statement within the context manager block could raise an exception, and control could resume after the context manager block. That is, it must assume that you could fall off the bottom of the function and implicitly return `None`.

Yeah I realize that now. Sorry for the noise.

On a side note, our code base makes extensive use of `ExitStack`. I am gradually adding type hints and this is a bit of a nuisance. I wish there were a version of `ExitStack` that did not provide the possibility of swallowing exceptions. But even if it existed, I am not sure it could be type-annotated in such a way that my example did not cause a warning. The behavior seemd cooked in the `with` statement. Although, if I replace the `ExitStack` with an `open`, I do not get the warning.

**Comment by JelleZijlstra** [comment]
This is affected by the return annotation of `__exit__`: https://mypy.readthedocs.io/en/stable/error_code_list.html?highlight=__Exit__#check-the-return-type-of-exit-exit-return

`ExitStack.__exit__` is annotated as returning bool: https://github.com/python/typeshed/blob/master/stdlib/contextlib.pyi#L75 That is correct because ExitStack can sometimes suppress exceptions.

To fix the error, you could subclass ExitStack and override the return type of `__exit__` to be None, although this would be unsafe if you do use ExitStack's error-suppression capabilities.

**Comment by jll63** [comment]
> To fix the error, you could subclass ExitStack and override the return type of `__exit__` to be None, although this would be unsafe if you do use ExitStack's error-suppression capabilities.

In that case, I guess would also override `enter_context` to accept only `ContextManager`s whose `__exit__` also returns `None`. But subclassing and overriding in this way violates the LSP, so a similar, but independent, mechanism would be better. I wonder if that already exists somewhere. I may implement this as an exercise otherwise. TBH, I find the practice of suppressing exceptions in `ExitStack` of questionable value (not talking about `contextmanager.suppress` here, which is obviously a useful thing).
